File: UIKit and Animations/WWDC 2014/A Look Inside Presentation Controllers (With a case study on alerts and search)/Sample_Codes/LookInside Presentation Controllers, Adaptivity, and Custom Animator Objects/LookInsidePresentationControllersAdaptivityandCustomAnimatorObjects/LookInside/AAPLRootViewController.m
/*
 Copyright (C) 2014 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 
  AAPLRootViewController implementation.
  
 */

#import "AAPLRootViewController.h"
#import "AAPLPhotoCollectionViewCell.h"
#import "AAPLOverlayViewController.h"

#import "AAPLOverlayTransitioner.h"
#import "AAPLCoolTransitioner.h"

#define kNumberOfViews (37)
#define kViewsWide (5)
#define kViewMargin (2.0)
#define kCellReuseIdentifier @"CellReuseIdentifier"

@implementation AAPLRootViewController

- (instancetype)init
{
    UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc] init];
    self = [super initWithCollectionViewLayout:layout];
    if (self)
    {
        [self configureTitleBar];
    }
    return self;
}

- (void)viewDidLoad
{
    [[self collectionView] registerClass:[AAPLPhotoCollectionViewCell class] forCellWithReuseIdentifier:kCellReuseIdentifier];
    [[self collectionView] setBackgroundColor:nil];
    UICollectionViewFlowLayout *layout = (UICollectionViewFlowLayout *)[self collectionViewLayout];
    
    [layout setMinimumInteritemSpacing:kViewMargin];
    [layout setMinimumLineSpacing:kViewMargin];
    
    [self viewWillTransitionToSize:[[self view] bounds].size withTransitionCoordinator:nil];
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    return kNumberOfViews;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    AAPLPhotoCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:kCellReuseIdentifier forIndexPath:indexPath];
    NSString *photoName = [@([indexPath item]) stringValue];
    UIImage *photo = [UIImage imageNamed:photoName];
    
    [cell setImage:photo];
    
    return cell;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    AAPLOverlayViewController *overlay = [[AAPLOverlayViewController alloc] init];
    
    if([self presentationShouldBeAwesome])
    {
        transitioningDelegate = [[AAPLCoolTransitioningDelegate alloc] init];
    }
    else
    {
        transitioningDelegate = [[AAPLOverlayTransitioningDelegate alloc] init];
    }

    [overlay setTransitioningDelegate:transitioningDelegate];

    AAPLPhotoCollectionViewCell *selectedCell = (AAPLPhotoCollectionViewCell *)[collectionView cellForItemAtIndexPath:indexPath];
    
    [overlay setPhotoView:selectedCell];
    
    [self presentViewController:overlay animated:YES completion:NULL];
}

- (BOOL)presentationShouldBeAwesome
{
    return [coolSwitch isOn];
}

- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator
{
    [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
    
    CGFloat itemWidth = size.width / kViewsWide;
    itemWidth -= kViewMargin;
    
    [(UICollectionViewFlowLayout *)[[self collectionView] collectionViewLayout] setItemSize:CGSizeMake(itemWidth, itemWidth)];
    [[self collectionViewLayout] invalidateLayout];
}

- (NSUInteger)supportedInterfaceOrientations
{
    return UIInterfaceOrientationMaskAll;
}

- (void)configureTitleBar
{
    [self setTitle:NSLocalizedString(@"LookInside Photos", @"App Title")];
    [self setEdgesForExtendedLayout:UIRectEdgeLeft | UIRectEdgeBottom | UIRectEdgeRight];
        
    coolSwitch = [[UISwitch alloc] init];
    [coolSwitch setOnTintColor:[UIColor purpleColor]];
    [coolSwitch setTintColor:[UIColor colorWithRed:1.0 green:0.0 blue:0.0 alpha:0.2]];
    
    UIBarButtonItem *enablecoolBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:coolSwitch];
    
    [[self navigationItem] setLeftBarButtonItem:enablecoolBarButtonItem];
}

@end

File: UIKit and Animations/WWDC 2014/A Look Inside Presentation Controllers (With a case study on alerts and search)/Sample_Codes/LookInside Presentation Controllers, Adaptivity, and Custom Animator Objects/LookInside_Swift/LookInside_Swift/AAPLRootViewController.swift
//
//  AAPLRootViewController.swift
//  LookInside_Swift
//
//  Created by Seyed Samad Gholamzadeh on 7/18/18.
//  Copyright © 2018 Seyed Samad Gholamzadeh. All rights reserved.
//

import UIKit

private let kNumberOfViews: Int = (37)
private let kViewsWide: CGFloat = (5)
private let kViewMargin: CGFloat = (2.0)

private let reuseIdentifier = "Cell"

class AAPLRootViewController: UICollectionViewController, UIAdaptivePresentationControllerDelegate {
	
	var coolSwitch: UISwitch!
	
	var transitionDelegate: UIViewControllerTransitioningDelegate!
	
	var presentationShouldBeAwesome: Bool {
		return self.coolSwitch.isOn
	}
	
	init() {
		let layout = UICollectionViewFlowLayout()
		super.init(collectionViewLayout: layout)
		self.configureTitleBar()
	}
	
	required init?(coder aDecoder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
	
	
	override func viewDidLoad() {
		super.viewDidLoad()
		
		self.collectionView!.register(AAPLPhotoCollectionViewCell.self, forCellWithReuseIdentifier: reuseIdentifier)
		self.collectionView?.backgroundColor = nil
		
		let layout = self.collectionViewLayout as! UICollectionViewFlowLayout
		layout.minimumInteritemSpacing = kViewMargin
		layout.minimumLineSpacing = kViewMargin
		self.configureLayouItemSize()
	}
	
	
	// MARK: UICollectionViewDataSource
	
	override func numberOfSections(in collectionView: UICollectionView) -> Int {
		// #warning Incomplete implementation, return the number of sections
		return 1
	}
	
	
	override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
		// #warning Incomplete implementation, return the number of items
		return kNumberOfViews
	}
	
	override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
		let cell = collectionView.dequeueReusableCell(withReuseIdentifier: reuseIdentifier, for: indexPath) as! AAPLPhotoCollectionViewCell
		let photoName = "\(indexPath.item)"
		let photo = UIImage(named: photoName)
		cell.image = photo
		return cell
	}
	
	// MARK: UICollectionViewDelegate
	override func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
		let overlay = AAPLOverlayViewController()
		
		if self.presentationShouldBeAwesome {
			self.transitionDelegate = AAPLCoolTransitioningDelegate()
		}
		else {
			self.transitionDelegate = AAPLOverlayTransitioningDelegate()
		}
		
		overlay.transitioningDelegate = self.transitionDelegate
		
		let pc = overlay.presentationController
		pc!.delegate = self
		
		let selectedCell = collectionView.cellForItem(at: indexPath) as! AAPLPhotoCollectionViewCell
		overlay.photoView = selectedCell
		self.present(overlay, animated: true, completion: nil)
	}
	
	override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
		super.viewWillTransition(to: size, with: coordinator)
		
		self.configureLayouItemSize(for: size)
	}
		
	func configureLayouItemSize(for size: CGSize? = nil) {
		let size = size ?? self.view.bounds.size
		var itemWidth = size.width / kViewsWide
		itemWidth -= kViewMargin
		(self.collectionViewLayout as! UICollectionViewFlowLayout).itemSize = CGSize(width: itemWidth, height: itemWidth)
		self.collectionViewLayout.invalidateLayout()
	}
	
	override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
		return .all
	}
	
	func configureTitleBar() {
		self.title = NSLocalizedString("LookInside Photos", comment: "App Title")
		self.edgesForExtendedLayout = UIRectEdge(rawValue: UIRectEdge.left.rawValue | UIRectEdge.bottom.rawValue | UIRectEdge.right.rawValue)
		self.coolSwitch = UISwitch()
		self.coolSwitch.onTintColor = .purple
		self.coolSwitch.tintColor = UIColor(red: 1.0, green: 0.0, blue: 0.0, alpha: 0.2)
		
		let enablecoolBarButtonItem = UIBarButtonItem(customView: self.coolSwitch)
		self.navigationItem.leftBarButtonItem = enablecoolBarButtonItem
	}
	
	func adaptivePresentationStyle(for controller: UIPresentationController) -> UIModalPresentationStyle {
		return .overFullScreen
	}
	
	
}
